typescript/src/redemption.ts

```diff
@@ -406,3 +410,103 @@ export async function getRedemptionRequest(
 
   return redemptionRequests[0]
 }
+
+/**
+ * Finds the oldest active wallet that has enough BTC to handle a redemption request.
+ * @param amount The amount to be redeemed in satoshis.
+ * @param bridge The handle to the Bridge on-chain contract.
+ * @param bitcoinClient Bitcoin client used to interact with the network.
+ * @param bitcoinNetwork Bitcoin network.
+ * @returns Promise with the wallet details needed to request a redemption.
+ */
+export async function findWalletForRedemption(
+  amount: BigNumber,
+  bridge: Bridge,
+  bitcoinClient: BitcoinClient,
+  bitcoinNetwork: BitcoinNetwork
+): Promise<{
+  walletPublicKey: string
+  mainUtxo: UnspentTransactionOutput
+}> {
+  const wallets = await bridge.getNewWalletRegisteredEvents()
+
+  let walletData:
+    | {
+        walletPublicKey: string
+        mainUtxo: UnspentTransactionOutput
+      }
+    | undefined = undefined
+  let maxAmount = BigNumber.from(0)
+
+  for (const wallet of wallets) {
+    const { state, mainUtxoHash, walletPublicKey } = await bridge.wallets(
+      wallet.walletPublicKeyHash
+    )
+
+    // Wallet must be in Live state.
+    if (
+      state !== WalletState.Live ||
+      mainUtxoHash.equals(
+        Hex.from(
+          "0x0000000000000000000000000000000000000000000000000000000000000000"
+        )
+      )
+    ) {
+      console.debug(
+        `Wallet is not in Live state or main utxo is not set. \
+        Continue the loop execution to the next wallet...`
+      )
+      continue
+    }
+
+    const walletBitcoinAddress = encodeToBitcoinAddress(
+      wallet.walletPublicKeyHash.toString(),
+      true,
+      bitcoinNetwork
+    )
+
+    const utxos = await bitcoinClient.findAllUnspentTransactionOutputs(
+      walletBitcoinAddress
+    )
+
+    // We need to find correct utxo- utxo components must point to the recent
+    // main UTXO of the given wallet, as currently known on the chain.
+    const mainUtxo = utxos.find((utxo) =>
+      mainUtxoHash.equals(bridge.buildUtxoHash(utxo))
+    )
```

Comment:
In case a wallet is working on something (e.g. redemption) and a bitcoin transaction was already submitted by the wallet to the bitcoin chain (new utxo returned from bitcoin client), but proof hasn't been submitted yet to the Bridge (old main utxo returned from the Bridge) this script will not find such a wallet.

cc @pdyraga

Reply:
There is another problem with the approach currently presented in `redemption.ts`: we completely ignore the value [of pending redemptions](https://github.com/keep-network/tbtc-v2/blob/f5f300b868137b6e8075eb69ca7e11f19143a912/solidity/contracts/bridge/Redemption.sol#L141). Say the wallet has 100 BTC, redemptions are processed every 3 hours, and so far, requests for 93 BTC has been registered in the Bridge. Based on the current version of the code, that wallet would still be selected for all redemptions <= 100 BTC and the transaction to Ethereum requesting redemption higher than 7 BTC would fail.

I am not sure why do we need to `findAllUnspentTransactionOutputs` and look for main UTXOs.

Why don't we get the main UTXO from the Bridge, take the value of all pending redemptions, subtract one from another, and use this value to determine if the given wallet is the right choice or not? This would be simpler and wouldn't have the two problems mentioned.

I think we can trust the Ethereum view of the world and not care about Bitcoin view of the world in this function.

Reply:
@pdyraga is right but:

> Why don't we get the main UTXO from the Bridge

This is not trivial as the main UTXO is stored as hash. To get it we have two options:
- Rely on contract events from transactions that change the main UXTO (so deposit sweeps, redemptions, moving funds, moved funds sweeps)
- Use similar logic than in https://github.com/keep-network/keep-core/blob/db145b629c3b4ec2677bc4457478439c685a2eff/pkg/tbtc/wallet.go#L392C6-L392C29 which needs to talk with Bitcoin but is simpler

Nevertheless, we must take the value of pending redemptions into account.

Reply:
9a0f289